src/api/mod.rs

```diff
@@ -1,27 +1,43 @@
+pub use self::cdtime::{nanos_to_collectd, CdTime};
+pub use self::logger::{collectd_log, log_err, CollectdLoggerBuilder, LogLevel};
+pub use self::oconfig::{ConfigItem, ConfigValue};
 use crate::bindings::{
-    data_set_t, hostname_g, plugin_dispatch_values, uc_get_rate, value_list_t, value_t, ARR_LENGTH,
-    DS_TYPE_ABSOLUTE, DS_TYPE_COUNTER, DS_TYPE_DERIVE, DS_TYPE_GAUGE,
+    data_set_t, hostname_g, meta_data_add_boolean, meta_data_add_double, meta_data_add_signed_int,
+    meta_data_add_string, meta_data_add_unsigned_int, meta_data_create, meta_data_destroy,
+    meta_data_get_boolean, meta_data_get_double, meta_data_get_signed_int, meta_data_get_string,
+    meta_data_get_unsigned_int, meta_data_t, meta_data_toc, meta_data_type, plugin_dispatch_values,
+    uc_get_rate, value_list_t, value_t, ARR_LENGTH, DS_TYPE_ABSOLUTE, DS_TYPE_COUNTER,
+    DS_TYPE_DERIVE, DS_TYPE_GAUGE, MD_TYPE_BOOLEAN, MD_TYPE_DOUBLE, MD_TYPE_SIGNED_INT,
+    MD_TYPE_STRING, MD_TYPE_UNSIGNED_INT,
 };
 use crate::errors::{ArrayError, CacheRateError, ReceiveError, SubmitError};
 use chrono::prelude::*;
 use chrono::Duration;
 use memchr::memchr;
 use std::borrow::Cow;
-use std::ffi::CStr;
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
 use std::fmt;
-use std::os::raw::c_char;
+use std::os::raw::{c_char, c_void};
 use std::ptr;
 use std::slice;
 use std::str::Utf8Error;
 
-pub use self::cdtime::{nanos_to_collectd, CdTime};
-pub use self::logger::{collectd_log, log_err, CollectdLoggerBuilder, LogLevel};
-pub use self::oconfig::{ConfigItem, ConfigValue};
-
 mod cdtime;
 mod logger;
 mod oconfig;
 
+/// The value of a metadata entry associated with a [ValueList].
+/// Metadata can be added using [ValueListBuilder::metadata] method.
+#[derive(Debug, Clone, PartialEq)]
+pub enum MetaValue {
```

Comment:
(thinking out loud), I'm assuming the reason why `MetaValue` is structured as owning the string instead of a string slice is because if we wanted:

```rust
pub enum MetaValue<'a> {
    String(&'a str),
    // ...
}
```

We'd still need an owned version for converting collectd's metadata to the hashmap due to the temporary nature of `meta_data_toc`, and it may be a bit premature to create two types: `SubmitMetaValue<'a>` and `MetaValue(Owned)`

Reply:
yes we would have to keep `meta_data_to` alive as long as the returned ValueList, probably with adding it as a field.

I don't have any experience with this type, but maybe std::borrow::Cow could help ?

Reply:
Sorry I've been afk. Yes Cow could help at the cost of some verbosity (caller would need to do something like `MetaValue::String(Cow::Borrowed("abc"))`). Overall I'm not too concerned and will defer to you if you think the string allocations are a cause for concern (I'm fine with changing the API as needed in the future).

I think this PR is just about done. Let me know if you have any changes queued up and then I'll give one final review.

Reply:
I agréé with you. I do not have any remaining changes to push. 
Ready for final review ! 😁

Reply:
Excellent, I'm away right now, but I should get back to this within a week and then kickoff the next release soon after